File: kbmtl_semisupervised_classification_variational_test.m
% Mehmet Gonen (mehmet.gonen@gmail.com)

function prediction = kbmtl_semisupervised_classification_variational_test(K, state)
    N = size(K, 2);
    T = size(state.W.mu, 2);

    prediction.H.mu = state.A.mu' * K;

    prediction.F.mu = prediction.H.mu' * state.W.mu;
    prediction.F.sigma = zeros(N, T);
    for t = 1:T
        prediction.F.sigma(:, t) = 1 + diag(prediction.H.mu' * state.W.sigma(:, :, t) * prediction.H.mu);
    end
    
    pos = 1 - normcdf((+state.parameters.margin - prediction.F.mu) ./ prediction.F.sigma);
    neg = normcdf((-state.parameters.margin - prediction.F.mu) ./ prediction.F.sigma);
    prediction.P = pos ./ (pos + neg);
end

File: kbmtl_semisupervised_classification_variational_test.R
# Mehmet Gonen (mehmet.gonen@gmail.com)

kbmtl_semisupervised_classification_variational_test <- function(K, state) {
  N <- dim(K)[2]
  T <- dim(state$W$mu)[2]

  H <- list(mu = crossprod(state$A$mu, K))

  F <- list(mu = crossprod(H$mu, state$W$mu), sigma = matrix(0, N, T))
  for (t in 1:T) {
    F$sigma[,t] <- 1 + diag(crossprod(H$mu, state$W$sigma[,,t]) %*% H$mu)
  }

  pos <- 1 - pnorm((+state$parameters$margin - F$mu) / F$sigma)
  neg <- pnorm((-state$parameters$margin - F$mu) / F$sigma)
  P <- pos / (pos + neg)

  prediction <- list(H = H, F = F, P = P)
}